File: day03/part2.py
#!/usr/bin/env python3
"""
aoc2021 solution skeleton
"""

import logging
import operator
from argparse import ArgumentParser
from pathlib import Path

import numpy as np


class Partnum:
    def __init__(self):
        self.number = ''
        self.points = []
        self.is_part = False

    def add_digit(self, val, point):
        self.number += val
        self.points.append(point)

    def __repr__(self):
        return f'Part({self.number}, {self.points})'


def get_adjacent_points(point, boundary):
    """uses row,col notation"""
    points = []

    for vector in [(-1,0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1), (-1, -1), (0, -1)]:
        tmp = tuple(map(operator.add, point, vector))
        if (tmp[0] < 0) or (tmp[0] > boundary[0]-1):
            continue
        if (tmp[1] < 0) or (tmp[1] > boundary[1]-1):
            continue
        points.append(tmp)

    return points


def is_symbol(char):
    if char.isdigit():
        return False
    if char == '.':
        return False
    return True


def main():
    """main"""

    parser = ArgumentParser()
    parser.add_argument('input')
    args = parser.parse_args()
    logging.info(args)

    data = Path(args.input).read_text('utf-8').splitlines()
    data = np.array(list(map(list, data)), dtype=str)
    rows, cols = data.shape

    print('data', data)
    print('shape', rows, cols)
    print('corners', data[0,0], data[-1, -1])

    parts = []
    part = Partnum()

    for r in range(rows):
        for c in range(cols):
            current = (r, c)

            if data[current].isdigit():
                part.add_digit(data[current], current)
                if any(is_symbol(data[x]) for x in get_adjacent_points(current, data.shape)):
                       part.is_part = True
                continue

            if not data[current].isdigit() and part.number:
                if part.is_part:
                    parts.append(part)
                part = Partnum()

        if part.number:
            if part.is_part:
                parts.append(part)
            part = Partnum()

    print(f'ans part1 {sum([int(x.number) for x in parts])}')
    print(parts)

    ans = 0

    for r in range(rows):
        for c in range(cols):
            current = (r, c)

            if data[current] == '*':  # is_gear()
                gear_parts = []

                for point in get_adjacent_points(current, data.shape):
                    for part in parts:
                        if point in part.points:
                            gear_parts.append(part)

                gear_parts = list(set(gear_parts))
                if len(gear_parts) == 2:
                    ans += int(gear_parts[0].number) * int(gear_parts[1].number)

                print(current, gear_parts)

    print(f'ans part2 {ans}')



if __name__ == '__main__':
    main()

File: day03/part2.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

// Partnum represents a structure to store number and points
type Partnum struct {
	number string
	points []point
	isPart bool
}

type point struct {
	row, col int
}

func (p *Partnum) addDigit(val string, row, col int) {
	p.number += val
	p.points = append(p.points, point{row, col})
}

func getAdjacentPoints(row, col int, rows, cols int) []point {
	points := []point{}

	directions := [8][2]int{
		{-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {-1, -1}, {0, -1},
	}

	for _, vector := range directions {
		tmpRow, tmpCol := row+vector[0], col+vector[1]

		if tmpRow < 0 || tmpRow >= rows || tmpCol < 0 || tmpCol >= cols {
			continue
		}

		points = append(points, point{tmpRow, tmpCol})
	}

	return points
}

func isSymbol(char string) bool {
	if _, err := strconv.Atoi(char); err == nil {
		return false
	}
	if char == "." {
		return false
	}
	return true
}

// isEqual checks if two Partnum structs are equal
func isEqual(a, b Partnum) bool {
	if a.number != b.number || a.isPart != b.isPart || len(a.points) != len(b.points) {
		return false
	}
	for i := range a.points {
		if a.points[i] != b.points[i] {
			return false
		}
	}
	return true
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run main.go <input_file>")
		return
	}

	inputFile := os.Args[1]

	file, err := os.Open(inputFile)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var data [][]string

	for scanner.Scan() {
		line := scanner.Text()
		row := []string{}
		for _, char := range line {
			row = append(row, string(char))
		}
		data = append(data, row)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	rows, cols := len(data), len(data[0])

	fmt.Println("data", data)
	fmt.Println("shape", rows, cols)
	fmt.Println("corners", data[0][0], data[rows-1][cols-1])

	parts := []Partnum{}
	part := Partnum{}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			current := point{r, c}

			if _, err := strconv.Atoi(data[current.row][current.col]); err == nil {
				part.addDigit(data[current.row][current.col], current.row, current.col)
				adjPoints := getAdjacentPoints(current.row, current.col, rows, cols)
				for _, x := range adjPoints {
					if isSymbol(data[x.row][x.col]) {
						part.isPart = true
						break
					}
				}
				continue
			}

			if !isSymbol(data[current.row][current.col]) && part.number != "" {
				if part.isPart {
					parts = append(parts, part)
				}
				part = Partnum{}
			}
		}

		if part.number != "" {
			if part.isPart {
				parts = append(parts, part)
			}
			part = Partnum{}
		}
	}

	sumParts := 0
	for _, x := range parts {
		num, _ := strconv.Atoi(x.number)
		sumParts += num
	}

	fmt.Println("ans part1", sumParts)
	fmt.Println(parts)

	ans := 0

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			current := point{r, c}

			if data[current.row][current.col] == "*" {
				gearParts := []Partnum{}

				adjPoints := getAdjacentPoints(current.row, current.col, rows, cols)
				for _, point := range adjPoints {
					for _, part := range parts {
						for _, p := range part.points {
							if point == p {
								gearParts = append(gearParts, part)
							}
						}
					}
				}

				uniqueGearParts := []Partnum{}
				for _, part := range gearParts {
					found := false
					for _, uniquePart := range uniqueGearParts {
						if isEqual(part, uniquePart) {
							found = true
							break
						}
					}
					if !found {
						uniqueGearParts = append(uniqueGearParts, part)
					}
				}g

				if len(uniqueGearParts) == 2 {
					num1, _ := strconv.Atoi(uniqueGearParts[0].number)
					num2, _ := strconv.Atoi(uniqueGearParts[1].number)
					ans += num1 * num2
				}

				fmt.Println(current, uniqueGearParts)
			}
		}
	}

	fmt.Println("ans part2", ans)
}
